lib/spack/spack/build_environment.py

```diff
@@ -525,22 +525,29 @@ def child_fun():
     carries on.
     """
 
-    def child_execution(child_connection):
+    def child_execution(child_connection, input_stream):
         try:
             setup_package(pkg, dirty=dirty)
-            function()
+            function(input_stream)
             child_connection.send([None, None, None])
         except Exception as e:
             child_connection.send([type(e), e, None])
         finally:
             child_connection.close()
 
     parent_connection, child_connection = multiprocessing.Pipe()
-    p = multiprocessing.Process(
-        target=child_execution,
-        args=(child_connection,)
-    )
-    p.start()
+    try:
+        # Forward sys.stdin to be able to activate / deactivate
+        # verbosity pressing a key at run-time
+        input_stream = os.fdopen(os.dup(sys.stdin.fileno()))
```

Comment:
@tgamblin A quote from [apue (3rd edition, chapter 8, section 3)](https://www.amazon.com/Advanced-Programming-UNIX-Environment-3rd/dp/0321637739):

> When we redirect the standard output of the parent from the program in Figure 8.1, the
child’s standard output is also redirected. Indeed, one characteristic of fork is that all
file descriptors that are open in the parent are duplicated in the child. We say
‘‘duplicated’’ because it’s as if the dup function had been called for each descriptor. The
parent and the child share a file table entry for every open descriptor (recall Figure 3.9).

So duplicating and forwarding it to the child seems to me the right thing to do to circumvent the fact that `multiprocessing` is closing the "effectively duplicated" `sys.stdin`. Underneath they will refer to the same file table entry and that's all we want.

Apologies if I restated the obvious for you, but as I don't usually program with such low level APIs I prefer to err on the side of repeating obvious things rather than not stating something that is not obvious at all. :smile: